### PR TITLE
Create core investments resource with history and operations

### LINKED ISSUE #17: Create Core Investment Resource
# Issue 17 — Create Core Investment Resource

**Description:**
Create the core Investment resource with Brazilian investment types, proper validation, and CRUD operations. This focuses purely on the data model and business logic without UI concerns. This issue also includes investment history and operation tracking for value changes.

**Acceptance Criteria:**

- [ ] Create `Moolah.Finance.Investment` Ash resource
- [ ] Add to `Moolah.Finance` domain
- [ ] Support Brazilian investment types with proper subtypes:
  - `renda_fixa`: cdb, lci_lca, cri_cra, debentures
  - `fundos`: renda_fixa, multimercado
  - `tesouro_direto`: selic, prefixado, ipca
  - `renda_variavel`: fiis, acoes
- [ ] Add required fields:
  - `name` (string, unique, required)
  - `type` (enum, required)
  - `subtype` (enum, required, validated against type)
  - `initial_value` (money, required)
  - `current_value` (money, required)
  - `account_id` (uuid, required) - references investment account
- [ ] Add optional fields:
  - `redemption_date` (date, optional)
  - `purchase_date` (date, optional)
- [ ] Create relationship to ledger account
- [ ] Add validation to ensure subtype matches investment type
- [ ] Add currency validation (investment currency must match account currency)
- [ ] Add account type validation (investment must link to `:investment_account`)
- [ ] Create database migration
- [ ] Implement CRUD actions (create, read, update, delete)
- [ ] Add unique constraint on investment names
- [ ] Write comprehensive tests for all validations and operations
- [ ] Track investment history records on create/update
- [ ] Track investment operations when value changes
- [ ] Add default read filter for active investments (include expired via alternate read action)

**Example Database Schema:**

Note: Feel free to use ecto or other types of migrations, rather than just writing pure sql

```sql
CREATE TABLE investments (
  id UUID PRIMARY KEY DEFAULT uuid_generate_v7(),
  name VARCHAR(255) NOT NULL UNIQUE,
  type VARCHAR(20) NOT NULL CHECK (type IN ('renda_fixa', 'fundos', 'tesouro_direto', 'renda_variavel')),
  subtype VARCHAR(20) NOT NULL CHECK (subtype IN ('cdb', 'lci_lca', 'cri_cra', 'debentures', 'renda_fixa', 'multimercado', 'selic', 'prefixado', 'ipca', 'fiis', 'acoes')),

  -- Using ash_money for proper currency handling
  initial_value MONEY_WITH_CURRENCY NOT NULL,
  current_value MONEY_WITH_CURRENCY NOT NULL,

  redemption_date DATE,
  purchase_date DATE,

  -- Link to the investment account in the ledger
  account_id UUID NOT NULL REFERENCES ledger_accounts(id),

  inserted_at TIMESTAMP WITH TIME ZONE NOT NULL DEFAULT NOW(),
  updated_at TIMESTAMP WITH TIME ZONE NOT NULL DEFAULT NOW(),

  -- Database-level validation for type/subtype combinations
  CHECK (
    (type = 'renda_fixa' AND subtype IN ('cdb', 'lci_lca', 'cri_cra', 'debentures')) OR
    (type = 'fundos' AND subtype IN ('renda_fixa', 'multimercado')) OR
    (type = 'tesouro_direto' AND subtype IN ('selic', 'prefixado', 'ipca')) OR
    (type = 'renda_variavel' AND subtype IN ('fiis', 'acoes'))
  )
);

-- Indexes for efficient querying
CREATE INDEX idx_investments_account_id ON investments(account_id);
CREATE INDEX idx_investments_type ON investments(type);
CREATE INDEX idx_investments_subtype ON investments(subtype);
CREATE UNIQUE INDEX idx_investments_unique_name ON investments(name);
```

**Suggested Ash Resource Structure:**

```elixir
defmodule Moolah.Finance.Investment do
  use Ash.Resource,
    domain: Moolah.Finance,
    data_layer: AshPostgres.DataLayer

  @types [:renda_fixa, :fundos, :tesouro_direto, :renda_variavel]
  @renda_fixa_subtypes [:cdb, :lci_lca, :cri_cra, :debentures]
  @fundos_subtypes [:renda_fixa, :multimercado]
  @tesouro_direto_subtypes [:selic, :prefixado, :ipca]
  @renda_variavel_subtypes [:fiis, :acoes]
  @all_subtypes @renda_fixa_subtypes ++ @fundos_subtypes ++
                @tesouro_direto_subtypes ++ @renda_variavel_subtypes

  postgres do
    table "investments"
    repo Moolah.Repo
  end

  actions do
    defaults [:read]

    create :create do
      accept [:name, :type, :subtype, :initial_value, :current_value,
              :redemption_date, :purchase_date, :account_id]

      validate {Moolah.Finance.Validations.ValidateSubtype, []}
      validate {Moolah.Finance.Validations.ValidateInvestmentCurrency, []}
      validate {Moolah.Finance.Validations.ValidateInvestmentAccountType, []}
    end

    update :update do
      accept [:name, :current_value, :redemption_date, :purchase_date]

      validate {Moolah.Finance.Validations.ValidateInvestmentCurrency, []}
      validate {Moolah.Finance.Validations.ValidateInvestmentAccountType, []}
    end

    destroy :destroy
  end

  attributes do
    uuid_v7_primary_key :id

    attribute :name, :string do
      allow_nil? false
      constraints max_length: 255, min_length: 1
    end

    attribute :type, :atom do
      constraints one_of: @types
      allow_nil? false
    end

    attribute :subtype, :atom do
      constraints one_of: @all_subtypes
      allow_nil? false
    end

    attribute :initial_value, :money do
      allow_nil? false
      constraints storage_type: :money_with_currency
    end

    attribute :current_value, :money do
      allow_nil? false
      constraints storage_type: :money_with_currency
    end

    attribute :redemption_date, :date
    attribute :purchase_date, :date

    timestamps()
  end

  relationships do
    belongs_to :account, Moolah.Ledger.Account do
      attribute_writable? true
      allow_nil? false
    end
  end

  identities do
    identity :unique_name, [:name]
  end
end
```

**Required Validation Modules:**

```elixir
# lib/moolah/finance/validations/validate_subtype.ex
defmodule Moolah.Finance.Validations.ValidateSubtype do
  use Ash.Resource.Validation

  @impl true
  def validate(changeset, _opts) do
    type = Ash.Changeset.get_attribute(changeset, :type)
    subtype = Ash.Changeset.get_attribute(changeset, :subtype)

    valid_subtypes = case type do
      :renda_fixa -> [:cdb, :lci_lca, :cri_cra, :debentures]
      :fundos -> [:renda_fixa, :multimercado]
      :tesouro_direto -> [:selic, :prefixado, :ipca]
      :renda_variavel -> [:fiis, :acoes]
      _ -> []
    end

    if subtype in valid_subtypes do
      :ok
    else
      {:error, field: :subtype, message: "is invalid for investment type #{type}"}
    end
  end
end

# lib/moolah/finance/validations/validate_investment_currency.ex
defmodule Moolah.Finance.Validations.ValidateInvestmentCurrency do
  use Ash.Resource.Validation

  @impl true
  def validate(changeset, _opts) do
    account_id = Ash.Changeset.get_attribute(changeset, :account_id)
    initial_value = Ash.Changeset.get_attribute(changeset, :initial_value)
    current_value = Ash.Changeset.get_attribute(changeset, :current_value)

    with {:ok, account} <- Moolah.Ledger.Account |> Ash.get(account_id),
         {:ok, initial_currency} <- get_currency(initial_value),
         {:ok, current_currency} <- get_currency(current_value) do

      cond do
        account.currency != initial_currency ->
          {:error, field: :initial_value, message: "currency must match account currency (#{account.currency})"}
        account.currency != current_currency ->
          {:error, field: :current_value, message: "currency must match account currency (#{account.currency})"}
        initial_currency != current_currency ->
          {:error, field: :current_value, message: "currency must match initial value currency"}
        true ->
          :ok
      end
    else
      _ -> {:error, field: :account_id, message: "invalid account"}
    end
  end

  defp get_currency(%Money{currency: currency}), do: {:ok, to_string(currency)}
  defp get_currency(_), do: {:error, :invalid_money}
end
```

**Domain Integration:**

```elixir
# Add to lib/moolah/finance.ex
defmodule Moolah.Finance do
  use Ash.Domain,
    otp_app: :moolah

  resources do
    resource Moolah.Finance.BudgetCategory
    resource Moolah.Finance.LifeAreaCategory
    resource Moolah.Finance.Tag
    resource Moolah.Finance.Investment  # Add this line
    # ... other resources
  end
end
```

**Test Coverage Requirements:**

- Investment creation with all type/subtype combinations
- Validation failures for invalid type/subtype combinations
- Currency validation (investment currency must match account currency)
- Account type validation (must be investment account)
- Unique name constraint
- CRUD operations
- History creation on create/update
- Operation creation on value change
- Active read filter behavior
- Edge cases (missing fields, invalid references, etc.)

**Technical Notes:**

- Uses ash_money for proper currency handling
- Database constraints enforce type/subtype relationships
- Calculations provide real-time profit/loss without storing redundant data
- Account relationship ensures investments are properly linked to ledger
- Future UI integration will build on this foundation

### PR DESCRIPTION
**Pull request by JuliaMathias**
# Create core investments resource with history and operations

Closes #17.

## Summary

This PR introduces the core Investment data model for Brazilian investment types, along with history snapshots and operation tracking to support time-series reporting and auditability. It also tightens validation around subtype/type, currency alignment, and account type, and adds a focused test suite to lock the behavior in.

## What is inside

- New Ash resources for investments, investment history, and investment operations (with detailed module docs and examples).
- Changes that automatically capture history snapshots on create/update and record value-change operations.
- Validations that enforce subtype/type compatibility, currency alignment with the linked account, and investment-account restriction.
- Migration creating investment tables, indexes, and constraints.
- Domain wiring for the new resources.
- Comprehensive tests covering CRUD, validations, history/operations, and edge cases.
- Documentation and agent guidelines updated for consistent doc/spec usage and example conventions.

## Details by area

Data model and actions

- Investment resource with Brazilian investment types/subtypes, unique name identity, active-only read filter, and CRUD actions.
- InvestmentHistory resource for daily/value snapshots.
- InvestmentOperation resource for delta tracking on value changes.

### Business logic

- CreateInvestmentHistory change: creates purchase-date and today snapshots on create; appends snapshots on updates that change value.
- TrackInvestmentOperation change: records deltas whenever current_value changes.

### Validations

- Subtype/type compatibility.
- Currency alignment with the linked account (including safe fallback behavior on invalid money values).
- Account type restriction to investment_account.

### Database

- New investments/history/operations tables, indexes, and constraints for type/subtype integrity.

### Testing

- Coverage for CRUD, validation failures, read filters, history snapshot behavior, and operation tracking.
- Unit coverage for change modules edge paths.

## How to review

1. Start with the migration for schema structure.
2. Review the Investment resource and its changes/validations.
3. Scan history/operation resources for snapshot and delta behavior.
4. Walk through tests for coverage of the critical paths and edge cases.

## Testing

- mix precommit
- MIX_ENV=test mix coveralls.detail --filter lib/moolah/finance/changes/track_investment_operation.ex
- MIX_ENV=test mix coveralls.detail --filter lib/moolah/finance/validations/validate_investment_currency.ex

### CODE REVIEW

File: priv/repo/migrations/20260203170000_add_investments_and_tracking.exs [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+    create index(:investments, [:subtype])
+
+    create constraint(:investments, :investments_type_subtype_check,
+             check: """
+               (type = 'renda_fixa' AND subtype IN ('cdb', 'lci_lca', 'cri_cra', 'debentures')) OR
+               (type = 'fundos' AND subtype IN ('renda_fixa', 'multimercado')) OR
+               (type = 'tesouro_direto' AND subtype IN ('selic', 'prefixado', 'ipca')) OR
+               (type = 'renda_variavel' AND subtype IN ('fiis', 'acoes'))
+             """
+           )
```
[the remark this answers is not in the document]
Comment (JuliaMathias) [reply priv/repo/migrations/20260203170000_add_investments_and_tracking.exs]: Added on_delete: :delete_all to both history and operation foreign keys so child rows are cleaned up with the investment.

File: lib/moolah/finance/changes/create_investment_history.ex [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+      with :ok <-
+             insert_history_entries(record.id, [
+               %{recorded_on: today, value: record.current_value}
+             ]) do
+        {:ok, record}
+      end
+    else
+      {:ok, record}
+    end
+  end
```
[the remark this answers is not in the document]
Comment (JuliaMathias) [reply lib/moolah/finance/changes/create_investment_history.ex]: Same‑day multiple snapshots are intentional; added a doc note to make that explicit.
